### ISSUE TITLE
Catch Exception when client cannot connect to backend

### ISSUE THREAD

**Issue by Yamakasi** [PHPSocialNetwork/phpfastcache#666]
**My question**

I can actually catch every exception which is needed but I was never be able to catch the exception when the cacheserver is not started so the client cannot connect. This happens with me on Couchbase but I'm not sure if this is the same for the other backends.

When crawling the code I got the idea that the Exception is too deep and not "bubbled up" by rethrowing so you never get it back in your getIntance() try/catch block.

This is my PHP errorlog on a connect when I turn off the couchbase server:

`[Sun Jan 06 22:06:01.025382 2019] [proxy_fcgi:error] [pid 2195] [client 192.168.255.1:51321] AH01071: Got error 'tion L:156 I:-257092588) <127.0.0.1:11210> (SOCK=38d47c1ef1189a7d) Failed to establish connection: LCB_NETWORK_ERROR (0x10), os errno=111\nPHP message: [cb,EROR] (server L:777 I:-257092588) <NOHOST:NOPORT> (CTX=(nil),,SRV=0x556fa9302530,IX=0) Connection attempt failed. Received LCB_NETWORK_ERROR (0x10) from libcouchbase, received 0 from operating system\nPHP message: [cb,WARN] (lcbio L:112 I:-257092588) Translating errno=111, lcb=0x2c to NETWORK_ERROR\nPHP message: [cb,EROR] (connection L:156 I:-257092588) <127.0.0.1:11210> (SOCK=35f5e8791fc4fa28) Failed to establish connection: LCB_NETWORK_ERROR (0x10), os errno=111\nPHP message: [cb,EROR] (server L:777 I:-257092588) <NOHOST:NOPORT> (CTX=(nil),,SRV=0x556fa9302530,IX=0) Connection attempt failed. Received LCB_NETWORK_ERROR (0x10) from libcouchbase, received 0 from operating system\nPHP message: [cb,WARN] (lcbio L:112 I:-257092588) Translating errno=111, lcb=0x2c to NETWORK_ERROR\nPHP message: [cb,EROR] (connection L:156 I:-257092588) <127.0.0.1:11210> (SOCK=db78bfd208b2b155) Failed to establish connection: LCB_NETWORK_ERROR (0x10), os errno=111\nPHP message: [cb,EROR] (server L:777 I:-257092588) <NOHOST:NOPORT> (CTX=(nil),,SRV=0x556fa9302530,IX=0) Connection attempt failed. Received LCB_NETWORK_ERROR (0x10) from libcouchbase, received 0 from operating system\nPHP message: [cb,WARN] (lcbio L:112 I:-257092588) Translating errno=111, lcb=0x2c to NETWORK_ERROR\nPHP message: [cb,EROR] (connection L:156 I:-257092588) <127.0.0.1:11210> (SOCK=f5f8df03dbdbb035) Failed to establish connection: LCB_NETWORK_ERROR (0x10), os errno=111\nPHP message: [cb,EROR] (server L:777 I:-257092588) <NOHOST:NOPORT> (CTX=(nil),,SRV=0x556fa9302530,IX=0) Connection attempt failed. Received LCB_NETWORK_ERROR (0x10) from libcouchbase, received 0 from operating system\nPHP message: [cb,WARN] (lcbio L:112 I:-257092588) Translating errno=111, lcb=0x2c to NETWORK_ERROR\nPHP message: [cb,EROR] (connection L:156 I:-257092588) <127.0.0.1:11210> (SOCK=7b9eb8f71c48e23f) Failed to establish connection: LCB_NETWORK_ERROR (0x10), os errno=111\nPHP message: [cb,EROR] (server L:777 I:-257092588) <NOHOST:NOPORT> (CTX=(nil),,SRV=0x556fa9302530,IX=0) Connection attempt failed. Received LCB_NETWORK_ERROR (0x10) from libcouchbase, received 0 from operating system\nPHP message: [cb,WARN] (retryq L:163 I:-257092588) Failing command (seq=332) from retry queue: LCB_NETWORK_ERROR (0x10)\nPHP message: [cb,WARN] (lcbio L:112 I:-257092588) Translating errno=111, lcb=0x2c to NETWORK_ERROR\nPHP message: [cb,EROR] (connection L:156 I:-257092588) <127.0.0.1:11210> (SOCK=bdbe97486e7d0b72) Failed to establish connection: LCB_NETWORK_ERROR (0x10), os errno=111\nPHP message: [cb,EROR] (server L:777 I:-257092588) <NOHOST:NOPORT> (CTX=(nil),,SRV=0x556fa9302530,IX=0) Connection attempt failed. Received LCB_NETWORK_ERROR (0x10) from libcouchbase, received 0 from operating system\nPHP message: [cb,WARN] (lcbio L:112 I:-257092588) Translating errno=111, lcb=0x2c to NETWORK_ERROR\nPHP message: [cb,EROR] (connection L:156 I:-257092588) <127.0.0.1:11210> (SOCK=56424032db6a1e68) Failed to establish connection: LCB_NETWORK_ERROR (0x10), os errno=111\nPHP message: [cb,EROR] (server L:777 I:-257092588) <NOHOST:NOPORT> (CTX=(nil),,SRV=0x556fa9302530,IX=0) Connection attempt failed. Received LCB_NETWORK_ERROR (0x10) from libcouchbase, received 0 from operating system\nPHP message: [cb,WARN] (lcbio L:112 I:-257092588) Translating errno=111, lcb=0x2c to NETWORK_ERROR\nPHP message: [cb,EROR] (connection L:156 I:-257092588) <127.0.0.1:11210> (SOCK=84a6124c2bff3250) Failed to establish connection: LCB_NETWORK_ERROR (0x10), os errno=111\nPHP message: [cb,EROR] (server L:777 I:-257092588) <NOHOST:NOPORT> (CTX=(nil),,SRV=0x556fa9302530,IX=0) Connection attempt failed. Received LCB_NETWORK_ERROR (0x10) from libcouchbase, received 0 from operating system`

**Comment by Geolim4** [comment]
Hello,

This error is not "catchable" since it's a issue of libcouchbase leading to an fcgi error therefore it's too low level to be catch in the php-level.

Cheers,
Georges.

**Comment by Yamakasi** [comment]
This is partly true, if you cannot open a bucket or perform an action on the backend you simply can throw an exception. This is what I did on my own couchbase implementation earlier as well, so I think, am pretty sure, you can (make a workaround).

**Comment by Geolim4** [comment]
Actually, whatever the driver, Phpfastcache only checks for extension/library/dependency availability.
I took the decision no to catch connection timeout/network/3rd party issues for many reasons:

- It's out of Phpfastcache's scope technically & theoretically
- I don't want to PFC becomes a labyrinthine system by catching everything that throw something
- Murphy's law:  "Anything that can go wrong will go wrong" meaning if something is wrongly configured or potentially unstable on the server, expect it to fail, not to be "handled by PFC".
- If an error like that occurs, you should not walks away from this and should take it seriously: The script should ends or initiate dying procedures and your intervention is required.

In summary: I won't fix that, not because I don't want to, but because I'd rather spend time improving features and code quality than spending time adding fallback/try-catch everywhere to "hide" hideous bugs from 3rd party drivers.

Have a good night !

**Comment by Yamakasi** [comment]
Hi.

Yes I agree, maybe we need to kick the couchbase boys a little bit, I know the maintainer and already pinged him to see how they see that.

It are difficult decisions to make sometimes in development like a library you have as you cannot make other software do it better then they already do to line them up with the rest (RFC way of thinking), undoable/too much "tards" in IT ;)

Thanks a lot!
